### DISCUSSION TITLE
Unsolved usecase: Soft-deletion with views

### DISCUSSION THREAD

**Discussion by Ploppz** [diesel-rs/diesel#3377] [Ideas]
I would like to share my use case. I'm trying to implement soft deletion on some or all tables in my project.
This article gives a good approach https://evilmartians.com/chronicles/soft-deletion-with-postgresql-but-with-logic-on-the-database
You don't necessarily have to read it. Basically: We create a *view* for each table which only selects the rows that are not soft-deleted.

Consequently, for SELECT statements we need to use the view, but for UPDATE/INSERT/DELETE statements I have had some problems using the view and instead in some cases I must use the underlying table itself.
The specific problem I have had happens with `INSERT ... ON CONFLICT ON CONSTRAINT ...`.
This means that I need to have the generated schema modules available for both the view and the table.

I got pretty far with a solution where I made every `table!` call evaluate into two `table!` calls, one for the view and one for the table (I have a naming scheme: the table is prefixed with `_`). It did work although it wasn't super pretty to have to care about exactly whether it's the view or the table modules that are in scope.

I did hit one problem that made me put this all aside: For every table, I generate a struct that acts as a filter, for example `ProjectFilter` for the `_project` table (`project` view).
This struct implements:
```rust
impl diesel::expression::AsExpression<diesel::sql_types::Nullable<diesel::sql_types::Bool>> for #filter_name {
    type Expression = Box<dyn #boxable_expr_type + 'static>;
```
where `#boxable_expr_type` is a type that involves the particular table that this filter works on.
The problem is that the generated `ProjectFilter` can only be used on _etiher_ the `project` view or the `_project` table. It's not possible to make it work on both afaik due to how associated types in traits works? But I need to be able to use it in any SELECT/INSERT/UPDATE/DELETE statement.

So that's where I'm stuck given that I chose to go this path with soft-deletion. 
What I think of as a possible solution:
Can we model views in `diesel` somehow that includes information about what is the underlying table? And make it possible to always in the Rust code use the _view_ for all operations SELECT/INSERT/DELETE/UPDATE, but behind-the-scenes we use the underlying table for INSERT/DELETE/UPDATE?

Otherwise, can we make it possible for a filter to be applicable to several tables?

Would like to hear any other suggestions and ideas.

**Comment by weiznich** [comment]
> The problem is that the generated ProjectFilter can only be used on etiher the project view or the _project table. It's not possible to make it work on both afaik due to how associated types in traits works? But I need to be able to use it in any SELECT/INSERT/UPDATE/DELETE statement.

It's not only a associated type problem. The inherent issue here is that you need to specify somehow which columns are used while building your filter. These columns always refer to a specific table, so either `_table` or `table` in your case. As these columns are represented by different types they are incompatible to each other, although they all implement the same expression types (but not `AppearsOnTable<…>`). That's the real problem you need to solve here.

One possible solution is to introduce a generic parameter in your `filter` type, which takes the table type (or whatever) as argument. This allows your to specialize the `AsExpression` implementations for specific arguments.

So you would have 
```rust
struct ProjectFilter<T> {
    // … your existing fields,
    p: PhantomData<T>, // does not hold any data
}

impl AsExpression<…> for ProjectFilter<__project> {
// …
}

impl AsExpression<…> for ProjectFilter<project> {
//…
}

```

> Can we model views in diesel somehow that includes information about what is the underlying table? And make it possible to always in the Rust code use the view for all operations SELECT/INSERT/DELETE/UPDATE, but behind-the-scenes we use the underlying table for INSERT/DELETE/UPDATE?

That's not possible in the general case, as views can reference an arbitrary number of tables. There is just no clear mapping between a view and tables, as a view is essentially a potentially really complex query (think CTE for example).

**Reply by Ploppz** [reply]
Thanks for swift reply!

>  The inherent issue here is that you need to specify somehow which columns are used while building your filter. 

I see. In the ideal (to me) scenario, I could use the same columns for the view and table but I understand that it's not possible.

> One possible solution is to introduce a generic parameter in your `filter` type, which takes the table type (or whatever) as argument.

Didn't think of that, thanks! It might become a bit cumbersome/unintuitive but sounds workable.

> That's not possible in the general case, as views can reference an arbitrary number of tables. There is just no clear mapping between a view and tables, as a view is essentially a potentially really complex query (think CTE for example).

What about looking at the column level? Like column X of the view comes from table X. I assume it's not possible to get this information but I wanted to ask anyway.

It's a conundrum that I'll keep in mind and see if I get any more ideas.

**Reply by weiznich** [reply]
> In the ideal (to me) scenario, I could use the same columns for the view and table but I understand that it's not possible.

Yes that's not possible.

>What about looking at the column level? Like column X of the view comes from table X. I assume it's not possible to get this information but I wanted to ask anyway.

Well that's really not possible here. The following view is also valid:

```SQL
CREATE VIEW test_view AS
WITH cte AS(
  SELECT id, name FROM users,
),
cte2 AS(
  SELECT id, "no name" FROM strange_users,
),
cte3 AS(
  SELECT id,first_middle + last AS name FROM some_other_users
)
  SELECT * FROM cte
  UNION
  SELECT * FROM cte2
  UNION 
  SELECT * FROM cte3;
```
(That's not even a very complicated view…)

to what exactly would the second column of that view map? It cannot map to a specific table, as results may come from different table + some tables may not even have a matching column and use a expression that refers to zero, one or multiple other columns there.
